src/main/scala/com/microsoft/ml/spark/stages/StratifiedRepartition.scala

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.stages
+
+import com.microsoft.ml.spark.core.contracts.{HasLabelCol, Wrappable}
+import org.apache.spark.RangePartitioner
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared.HasSeed
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+/** Constants for <code>StratifiedRepartition</code>. */
+object SPConstants {
+  val COUNT = "count"
+  val EQUAL = "equal"
+  val ORIGINAL = "original"
+}
+
+object StratifiedRepartition extends DefaultParamsReadable[DropColumns]
+
+/** <code>StratifiedRepartition</code> repartitions the DataFrame such that each label is selected in each partition.
+  * This may be necessary in some cases such as in LightGBM multiclass classification, where it is necessary for
+  * at least one instance of each label to be present on each partition.
+  */
+class StratifiedRepartition(val uid: String) extends Transformer with Wrappable
+  with DefaultParamsWritable with HasLabelCol with HasSeed {
+  def this() = this(Identifiable.randomUID("StratifiedRepartition"))
+
+  val fraction = new Param[String](this, "fraction",
+    "Specify equal to repartition with replacement across all labels, otherwise specify " +
+      "original to keep the ratios in the original dataset")
+  setDefault(fraction -> "equal")
+
+  def getFraction: String = $(fraction)
+  def setFraction(value: String): this.type = set(fraction, value)
+
+  /** @param dataset - The input dataset, to be transformed
+    * @return The DataFrame that results from stratified repartitioning
+    */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    // Count unique values in label column
+    val distinctLabelCounts = dataset.select(getLabelCol).groupBy(getLabelCol).count().collect()
+
+    val labelToCount = distinctLabelCounts.map(row => (row.getInt(row.schema.fieldIndex(getLabelCol)),
+      row.getLong(row.schema.fieldIndex(SPConstants.COUNT))))
+
+    val labelToFraction =
+      if (getFraction == SPConstants.EQUAL) {
+        val maxLabelCount = labelToCount.map(lc => lc._2).max
+        labelToCount.map(lc => (lc._1, maxLabelCount.toDouble / lc._2)).toMap
+      } else {
+        labelToCount.map(lc => (lc._1, 1.0)).toMap
+      }
+
+    val spdata = dataset.toDF().rdd.keyBy(row => row.getInt(row.schema.fieldIndex(getLabelCol)))
```

Comment:
Does this have any equivalent in data frame API?

Reply:
hmm, it seems not